### PR TITLE
Adding Cost function parameters 

### LINKED ISSUE #701: Issue with MarketBidCost when linking problems in simulation.
I’m trying to define a simulation in the following way to represent the UC (to solve for commitments) and EDs (to solve for prices) of a DA operation. However, the simulation is not updating correctly the ED values according to the time series defined for the MarketBidCost.

The simulation sequence:
```
problems = SimulationProblems(;
        UC=OperationsProblem(
            template_uc,
            system_uc;
            optimizer=solver_uc,
            system_to_file=system_to_file,
            kwargs...,
        ),
        ED=OperationsProblem(
            template_ed,
            system_ed;
            optimizer=solver_ed,
            system_to_file=system_to_file,
            services_slack_variables=services_slack_variables,
            balance_slack_variables=balance_slack_variables,
            constraint_duals=constraint_duals,
            kwargs...,
        ),
    )
    feedforward_chronologies = Dict(("UC" => "ED") => Synchronize(; periods=24))
    feedforward = Dict(
        ("ED", :devices, :ThermalStandard) =>
            SemiContinuousFF(; binary_source_problem=ON, affected_variables=[ACTIVE_POWER]),
    )
    intervals = Dict("UC" => (Hour(24), Consecutive()), "ED" => (Hour(1), Consecutive()))
    uc_ed_sequence = SimulationSequence(;
        problems=problems,
        intervals=intervals,
        ini_cond_chronology=InterProblemChronology(),
        feedforward_chronologies=feedforward_chronologies,
        feedforward=feedforward,
    )
```
The `MarketBidCost` "variable cost" SingleTimeSeries added to both UC and ED for generator `bus4_virtual_supply`:
image

The `MarketBidCost` "variable cost" Deterministic forecast in UC:
```
julia> get_time_series_values(Deterministic, gen_uc, "variable_cost")
24-element Vector{Float64}:
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 0.0
 
 
 julia> get_time_series(Deterministic, gen_uc, "variable_cost")
DeterministicSingleTimeSeries time_series (7):
   single_time_series: SingleTimeSeries time_series (168)
   initial_timestamp: 2020-01-01T00:00:00
   interval: 24 hours
   count: 7
   horizon: 24
```

The `MarketBidCost` "variable cost" Deterministic forecast in to ED:
```
julia> get_time_series_values(Deterministic, gen_ed, "variable_cost"; start_time=DateTime("2020-01-01")+Hour(23))
1-element Vector{Float64}:
 0.0


julia> get_time_series(Deterministic, gen_ed, "variable_cost"; start_time=DateTime("2020-01-01")+Hour(23))
DeterministicSingleTimeSeries time_series (145):
   single_time_series: SingleTimeSeries time_series (145)
   initial_timestamp: 2020-01-01T23:00:00
   interval: 1 hour
   count: 145
   horizon: 1
```

The ED problem for hour 24 is using the cost hour 1:
```
julia> print(sim.problems[:ED].internal.optimization_container.JuMPmodel)
Min 3000 P__ThermalStandard_{Solitude, 1} + 1500 P__ThermalStandard_{Park City, 1} + 1.0e9 P__ThermalStandard_{bus4_virtual_supply, 1} + 1400 P__ThermalStandard_{Alta, 1} + 1000 P__ThermalStandard_{Brighton, 1} + 4000 P__ThermalStandard_{Sundance, 1}
```

### PR DESCRIPTION
**Pull request by sourabhdalvi**
Fixes #701

### CODE REVIEW

File: src/core/optimization_container.jl [hunk 1]
```diff
@@ -171,6 +192,48 @@ get_initial_conditions_data(container::OptimizationContainer) =
...
+end
+
+function has_container_key(
+    container::OptimizationContainer,
+    ::Type{T},
+    ::Type{U},
+    meta = CONTAINER_KEY_EMPTY_META,
+) where {T <: ConstraintKey, U <: Union{PSY.Component, PSY.System}}
+    key = ConstraintKey(T, U, meta)
+    return haskey(container.variables, key)
```
Comment (daniel-thom) [inline src/core/optimization_container.jl:225]: `container.duals`?

File: src/devices_models/devices/common/cost_functions.jl [hunk 2]
```diff
@@ -232,33 +241,67 @@ function pwl_gencost_sos!(
...
-    if spec.sos_status == SOSStatusVariable.NO_VARIABLE
+    if sos_status == SOSStatusVariable.NO_VARIABLE
         bin = 1.0
         @debug "Using Piecewise Linear cost function but no variable/parameter ref for ON status is passed. Default status will be set to online (1.0)" _group =
             LOG_GROUP_COST_FUNCTIONS
 
-    elseif spec.sos_status == SOSStatusVariable.PARAMETER
-        param_key = encode_symbol("ON", string(spec.component_type))
+    elseif sos_status == SOSStatusVariable.PARAMETER
+        param_key = encode_symbol("ON", string(T))
```
Comment (daniel-thom) [inline src/devices_models/devices/common/cost_functions.jl:299]: Is this vulnerable to a bug if T has the module name, such as `PowerSystems.ThermalStandard`?

Comment (sourabhdalvi) [reply src/devices_models/devices/common/cost_functions.jl:299]: Should I use something like `encode_symbol("ON", string(IS.strip_module_name(T)))`

Comment (daniel-thom) [reply src/devices_models/devices/common/cost_functions.jl:299]: Can you create the actual ParameterKey? Maybe something like `ParameterKey(OnStatusParameter, T)`?

**Comment by jd-lara** [comment]
looks like the failing tests are due to SCIP initial conditions testing

File: src/parameters/update_parameters.jl [hunk 3]
```diff
@@ -226,3 +226,58 @@ function update_parameter_values!(
...
+                start_time = initial_forecast_time,
+                len = horizon,
+            )
+            variable_cost_forecast_values = TimeSeries.values(ts_vector)
+            for (t, value) in enumerate(variable_cost_forecast_values)
+                if attributes.uses_compact_power
+                    value, _ = _convert_variable_cost(value)
+                end
+                _set_param_value!(param_array, PSY.get_cost(value), name, t)
+                variable_cost!(container, param_array, attributes, component, t)
```
Comment (jd-lara) [inline src/parameters/update_parameters.jl:273]: This function is adding the pwl variables every time at every update call.

**Review by jd-lara — CHANGES_REQUESTED** [review]
This PR has a significant issue in that it adds again all the pwl variables at each parameter update call.

File: src/parameters/update_parameters.jl (hunk 3 above)
Comment (jd-lara) [reply src/parameters/update_parameters.jl]: fixed in the subsequent commits

File: src/core/variables.jl [hunk 4]
```diff
@@ -125,6 +125,9 @@ struct ComponentActivePowerReserveDownVariable <: SubComponentVariableType end
 # Necessary as a work around ofr HVDC models with losses
 struct HVDCTotalPowerDeliveredVariable <: VariableType end
 
+struct PieceWiseLinearCostVariable <: VariableType end
+
 const START_VARIABLES = (HotStartVariable, WarmStartVariable, ColdStartVariable)
 
 write_resulting_value(::Type{<:VariableType}) = true
+write_resulting_value(::Type{PieceWiseLinearCostVariable}) = false
```
Comment (daniel-thom) [inline src/core/variables.jl:133]: I don't recall where I commented before, but this function does not sound like a boolean.

File: src/devices_models/devices/common/set_expression.jl [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+Replaces an expression value in the expression container if the key exists
+"""
+function set_expression!(
+    container::OptimizationContainer,
+    ::Type{S},
+    cost_expression::JuMP.AbstractJuMPScalar,
+    component::T,
+    time_period::Int,
+) where {S <: CostExpressions, T <: PSY.Component}
+    if has_container_key(container, S, T)
```
Comment (daniel-thom) [inline src/devices_models/devices/common/set_expression.jl:11]: The behavior in the case where the key doesn't exist could be surprising to a user. Is that really the best thing to do? Should there be a log message at least?

File: src/core/optimization_container.jl [hunk 6]
```diff
@@ -48,6 +48,27 @@ function Base.isempty(pvc::PrimalValuesCache)
     return isempty(pvc.variables_cache) && isempty(pvc.expressions_cache)
 end
 
+mutable struct ObjectiveFunction
```
Comment (daniel-thom) [inline src/core/optimization_container.jl:51]: I don't see this being used or tested. Expected?

File: src/core/variables.jl (hunk 4 above)
Comment (jd-lara) [reply src/core/variables.jl:133]: This is a comment in the base PR. Will be fixed there.

File: src/core/optimization_container.jl (hunk 6 above)
Comment (jd-lara) [reply src/core/optimization_container.jl:51]: It is covered by all the uses of the optimization container.

**Pull request merged**